Analysis/Tasks/PWGCF/o2femtodream/include/o2femtodream/FemtoDreamObjectSelection.h

```diff
@@ -0,0 +1,156 @@
+// or submit itself to any jurisdiction.
+
+/// \file FemtoDreamObjectSelection.h
+/// \brief FemtoDreamObjectSelection - Partent class of all selections
+/// \author Andi Mathis, TU München, andreas.mathis@ph.tum.de
+
+#ifndef ANALYSIS_TASKS_PWGCF_FEMTODREAM_FEMTODREAMOBJECTSELECTION_H_
+#define ANALYSIS_TASKS_PWGCF_FEMTODREAM_FEMTODREAMOBJECTSELECTION_H_
+
+#include "FemtoDreamSelection.h"
+
+#include "ReconstructionDataFormats/PID.h"
+#include "Framework/HistogramRegistry.h"
+#include <Rtypes.h>
+
+#include <string>
+#include <cmath>
+
+using namespace o2::framework;
+
+namespace o2::analysis
+{
+namespace femtoDream
+{
+
+/// \class FemtoDreamObjectSelection
+/// \brief Cut class to contain and execute all cuts applied to tracks
+
+template <class T1, class T2>
+class FemtoDreamObjectSelection
+{
+ public:
+  virtual ~FemtoDreamObjectSelection() = default;
+
+  /// Initializes histograms for the task
+  void init(HistogramRegistry* registry)
+  {
+    mHistogramRegistry = registry;
+  }
+
+  void fillSelectionHistogram(const char* name)
+  {
+    int nBins = mSelections.size();
+    mHistogramRegistry->add(name, "; Cut; Value", kTH1F, {{nBins, 0, static_cast<double>(nBins)}});
+    auto hist = mHistogramRegistry->get<TH1>(HIST("TrackCuts/cuthist"));
+    for (size_t i = 0; i < mSelections.size(); ++i) {
+      hist->GetXaxis()->SetBinLabel(i + 1, Form("%u", mSelections.at(i).getSelectionVariable()));
+      hist->SetBinContent(i + 1, mSelections.at(i).getSelectionValue());
+    }
+  }
+
+  template <typename T>
+  void setSelection(T& selVals, T2 selVar, femtoDreamSelection::SelectionType selType)
+  {
+    std::vector<T1> tmpSelVals = selVals; // necessary due to some features of the Configurable
+    std::vector<FemtoDreamSelection<T1, T2>> tempVec;
+    for (const T1 selVal : tmpSelVals) {
+      tempVec.push_back(FemtoDreamSelection<T1, T2>(selVal, selVar, selType));
+    }
+    setSelection(tempVec);
+  }
+
+  void setSelection(std::vector<FemtoDreamSelection<T1, T2>>& sels)
+  {
+    switch (sels.at(0).getSelectionType()) {
+      case (femtoDreamSelection::SelectionType::kUpperLimit):
+      case (femtoDreamSelection::SelectionType::kAbsUpperLimit):
+        std::sort(sels.begin(), sels.end(), [](FemtoDreamSelection<T1, T2> a, FemtoDreamSelection<T1, T2> b) {
+          return a.getSelectionValue() > b.getSelectionValue();
+        });
+        break;
+      case (femtoDreamSelection::SelectionType::kLowerLimit):
+      case (femtoDreamSelection::SelectionType::kAbsLowerLimit):
+      case (femtoDreamSelection::SelectionType::kEqual):
+        std::sort(sels.begin(), sels.end(), [](FemtoDreamSelection<T1, T2> a, FemtoDreamSelection<T1, T2> b) {
+          return a.getSelectionValue() < b.getSelectionValue();
+        });
+        break;
+    }
+    for (const auto sel : sels) {
+      mSelections.push_back(sel);
+    }
+  }
+
+  T1 getMinimalSelection(T2 selVar, femtoDreamSelection::SelectionType selType)
+  {
+    T1 minimalSel;
```

Comment:
These will be executed per object? We may need to profile this? I wonder if instead one could produce a partition expression which does the selection instead of doing it by hand. @aalkin @saganatt what do you think?

Reply:
hmm, this is executed once per variable and run, so I don't think see why this should be a performance bottleneck

Reply:
That is then fine, but you also go through the switch for each track elsewhere, don't you?